README.md
# Bioplatforms Australia  Metadata Management

BPA Metadata is a web portal into Bioplatforms Australia's genomics sequencing data 
archive. It aims to aggregate all relevant meta and sequence data into a coherent data 
navigation system.

# Development
Ensure a late version of both docker and docker-compose are available in your environment.

bpa-metada is available as a fully contained Dockerized stack. The dockerised stack are used for both production
and development. Appropiate configuration files are available depending on usage.

Note that for data ingestion to work you need passwords to the hosted data, these are available from BPA on request.
Set passwords in your environment, these will be passed to the container.

## Quick Setup

* [Install docker and compose](https://docs.docker.com/compose/install/)
* git clone https://github.com/muccg/bpa-metadata.git
* `./develop.sh dev_build`

`develop.sh dev` will spin up the stack. See `./develop.sh usage` for some utility methods, which typically are simple 
wrappers arround docker and docker-compose.

docker-compose will fire up the stack like below:
```
docker ps -f name="bpam*"

IMAGE                       PORTS                                                                          NAMES
muccg/nginx-uwsgi:1.10      0.0.0.0:8080->80/tcp, 0.0.0.0:8443->443/tcp                                    bpametadata_nginx_1
mdillon/postgis:9.5         0.0.0.0:32944->5432/tcp                                                        bpametadata_db_1
memcached:1.4               11211/tcp                                                                      bpametadata_cache_1
muccg/bpametadata-dev       0.0.0.0:9000-9001->9000-9001/tcp, 8000/tcp, 0.0.0.0:9100-9101->9100-9101/tcp   bpametadata_uwsgi_1
muccg/bpametadata-dev       9000-9001/tcp, 0.0.0.0:8000->8000/tcp, 9100-9101/tcp                           bpametadata_runserver_1
```

Setup mirror config
```
docker exec -it bpametadata_runserver_1 /app/docker-entrypoint.sh django-admin set_mirrors
```

To execute the ingestion scripts run `docker exec -it bpametadata_runserver_1 /app/docker-entrypoint.sh django-admin`, be sure
to have the relevant variables below available in the environment. 

```
DJANGO_MAILGUN_API_KEY
BASE_REQUEST_LIST
BPA_MELANOMA_DOWNLOADS_PASSWORD
BPA_BASE_DOWNLOADS_PASSWORD
BPA_USERS_DOWNLOADS_PASSWORD
BPA_GBR_DOWNLOADS_PASSWORD
BPA_SEPSIS_DOWNLOADS_PASSWORD
```



```bash
./develop.sh

  [ INFO ] ./develop.sh 
  [  OK  ] Docker ip 172.17.0.1
  [ INFO ] http proxy
  [  OK  ] Proxy http://172.17.0.1:3128
  [  OK  ] HTTP proxy http://172.17.0.1:3128
  [ INFO ] pip proxy
  [  OK  ] Pip index url http://172.17.0.1:3141/root/pypi/+simple/
  [ INFO ] git tag
  [ INFO ] Ignoring tags, not on master branch
  [  OK  ] git tag: next_release

Environment:
 Pull during build              DOCKER_PULL                 1 
 No cache during build          DOCKER_NO_CACHE             0 
 Use proxy during builds        DOCKER_BUILD_PROXY          --build-arg http_proxy
 Push/pull from docker hub      DOCKER_USE_HUB              0
 Release docker image           DOCKER_IMAGE                muccg/bpametadata
 Use a http proxy               SET_HTTP_PROXY              1
 Use a pip proxy                SET_PIP_PROXY               1

Usage:
 ./develop.sh (baseimage|buildimage|devimage|releasetarball|prodimage)
 ./develop.sh (dev|dev_build)
 ./develop.sh (start_prod|prod_build)
 ./develop.sh (runtests|lettuce|selenium)
 ./develop.sh (start_test_stack|start_seleniumhub|start_seleniumtests|start_prodseleniumtests)
 ./develop.sh (pythonlint|jslint)
 ./develop.sh (ci_docker_staging|docker_staging_lettuce)
 ./develop.sh (ci_docker_login)

Example, start dev with no proxy and rebuild everything:
SET_PIP_PROXY=0 SET_HTTP_PROXY=0 ./develop.sh dev_rebuild


```

## Sites
- *Production* https://downloads.bioplatforms.com

## Licence
BPA Metadata is released under the GNU Affero GPL. See source for a licence copy.

## Contributing
* Fork next_release branch
* Make changes on a feature branch
* Submit pull request

## CCG Internal documentation

https://ccgmurdoch.atlassian.net/wiki/display/BM/BPA+Metadata+Home

[1]: Be aware that all CDN supported content still needs to be fetched from the internet. 


CHANGELOG.md
# Change Log

## [4.6.1](https://github.com/muccg/bpa-metadata/tree/4.6.1) (2016-12-21)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.6.0...4.6.1)

**Implemented enhancements:**

- Metagenomics protocol PDF for MM [\#123](https://github.com/muccg/bpa-metadata/issues/123)

## [4.6.0](https://github.com/muccg/bpa-metadata/tree/4.6.0) (2016-12-16)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.5.0...4.6.0)

**Implemented enhancements:**

- Deprecate sepsis tracker API - migrate fields to sample API [\#105](https://github.com/muccg/bpa-metadata/issues/105)
- migrate tracker ticket links to github issues [\#47](https://github.com/muccg/bpa-metadata/issues/47)
- Methods tab for MM [\#122](https://github.com/muccg/bpa-metadata/issues/122)
- Minor changes to sepsis webpage [\#121](https://github.com/muccg/bpa-metadata/issues/121)
- Remove legacy sepsis sample tracking view [\#119](https://github.com/muccg/bpa-metadata/issues/119)
- API: add support for proteomics swathms data [\#112](https://github.com/muccg/bpa-metadata/issues/112)
- API: add support for Metabalomics LCMS files [\#104](https://github.com/muccg/bpa-metadata/issues/104)
- Metatranscriptome support for marine microbes [\#99](https://github.com/muccg/bpa-metadata/issues/99)
- API: Add support for Proteomics Deep LC-MS files [\#98](https://github.com/muccg/bpa-metadata/issues/98)
- Remove the URL checked fields from all projects [\#91](https://github.com/muccg/bpa-metadata/issues/91)
- sepsis: import contextual metadata [\#81](https://github.com/muccg/bpa-metadata/issues/81)

**Fixed bugs:**

- -omics field incorrect in v3 API for Pacbio and Miseq data [\#120](https://github.com/muccg/bpa-metadata/issues/120)

## [4.5.0](https://github.com/muccg/bpa-metadata/tree/4.5.0) (2016-11-19)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.4.1...4.5.0)

**Implemented enhancements:**

- Update tracker names for Proteomics DeepLCMS, Metabolomics DeepLCMS [\#117](https://github.com/muccg/bpa-metadata/issues/117)
- API: Tracker fields update for Pacbio, Miseq [\#116](https://github.com/muccg/bpa-metadata/issues/116)

## [4.4.1](https://github.com/muccg/bpa-metadata/tree/4.4.1) (2016-11-15)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.4.0...4.4.1)

**Implemented enhancements:**

- rename sepsis to "Antibiotic Resistant Sepsis Pathogens" [\#115](https://github.com/muccg/bpa-metadata/issues/115)

## [4.4.0](https://github.com/muccg/bpa-metadata/tree/4.4.0) (2016-11-07)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.3.0...4.4.0)

**Implemented enhancements:**

- Admin - remove all tracker unrelated links from side menu under ARPI [\#113](https://github.com/muccg/bpa-metadata/issues/113)

**Fixed bugs:**

- BPA MM Project - Metagenome data imports [\#94](https://github.com/muccg/bpa-metadata/issues/94)
- bpam tracker: date format wrong [\#114](https://github.com/muccg/bpa-metadata/issues/114)
- ARPI: breadcrumb for transcriptomics hiseq is incorrect [\#111](https://github.com/muccg/bpa-metadata/issues/111)
- BASE access request: email address not sent in notification [\#109](https://github.com/muccg/bpa-metadata/issues/109)
- Sepsis: import of data\_type through admin is failing [\#108](https://github.com/muccg/bpa-metadata/issues/108)
- Sepsis tracker: show data as 'embargoed', not 'public' [\#107](https://github.com/muccg/bpa-metadata/issues/107)
- Missing miseq metadata from portal [\#88](https://github.com/muccg/bpa-metadata/issues/88)

## [4.3.0](https://github.com/muccg/bpa-metadata/tree/4.3.0) (2016-10-26)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.2.0...4.3.0)

**Implemented enhancements:**

- Marine Microbes: new tracker layout [\#100](https://github.com/muccg/bpa-metadata/issues/100)

**Fixed bugs:**

- Metadata ingest: float data in header causes stack trace [\#103](https://github.com/muccg/bpa-metadata/issues/103)

## [4.2.0](https://github.com/muccg/bpa-metadata/tree/4.2.0) (2016-10-24)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.1.1...4.2.0)

**Implemented enhancements:**

- API: Make all API endpoints read-only  [\#92](https://github.com/muccg/bpa-metadata/issues/92)
- Login screen needs attention [\#87](https://github.com/muccg/bpa-metadata/issues/87)
- remove TogetherJS from django-admin [\#86](https://github.com/muccg/bpa-metadata/issues/86)
- remove JIRA integration [\#85](https://github.com/muccg/bpa-metadata/issues/85)
- sepsis overview improvements [\#83](https://github.com/muccg/bpa-metadata/issues/83)
- sepsis tracker: add field for pilot data [\#82](https://github.com/muccg/bpa-metadata/issues/82)

**Fixed bugs:**

- review use of CDNs [\#27](https://github.com/muccg/bpa-metadata/issues/27)

## [4.1.1](https://github.com/muccg/bpa-metadata/tree/4.1.1) (2016-10-10)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.1.0...4.1.1)

**Implemented enhancements:**

- Add new / rename existing marine microbes datatypes [\#80](https://github.com/muccg/bpa-metadata/issues/80)
- Marine microbes contextual data updates [\#69](https://github.com/muccg/bpa-metadata/issues/69)

**Fixed bugs:**

- Marine microbes: transfer issues [\#70](https://github.com/muccg/bpa-metadata/issues/70)
- Degrees symbol in CSV spreadsheet breaks contextual import [\#67](https://github.com/muccg/bpa-metadata/issues/67)

## [4.1.0](https://github.com/muccg/bpa-metadata/tree/4.1.0) (2016-10-03)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.0.39...4.1.0)

**Implemented enhancements:**

- tracker: public view of project status for sepsis [\#72](https://github.com/muccg/bpa-metadata/issues/72)

**Fixed bugs:**

- traceback attempting CSV import [\#79](https://github.com/muccg/bpa-metadata/issues/79)

## [4.0.39](https://github.com/muccg/bpa-metadata/tree/4.0.39) (2016-09-30)
[Full Changelog](https://github.com/muccg/bpa-metadata/compare/4.0.38...4.0.39)

**Implemented enhancements:**


lib.sh
#!/bin/sh
#
# common definitons shared between projects
#
set -a

TOPDIR=$(cd `dirname $0`; pwd)
DATE=`date +%Y.%m.%d`

: ${DOCKER_BUILD_PROXY:="--build-arg http_proxy"}
: ${DOCKER_USE_HUB:="0"}
: ${DOCKER_IMAGE:="muccg/${PROJECT_NAME}"}
: ${SET_HTTP_PROXY:="1"}
: ${SET_PIP_PROXY:="1"}
: ${DOCKER_NO_CACHE:="0"}
: ${DOCKER_PULL:="1"}

# Do not set these, they are vars used below
DOCKER_ROUTE=''
DOCKER_BUILD_OPTS=''
DOCKER_RUN_OPTS='-e PIP_INDEX_URL -e PIP_TRUSTED_HOST'
DOCKER_COMPOSE_BUILD_OPTS=''


usage() {
    echo ""
    echo "Environment:"
    echo " Pull during build              DOCKER_PULL                 ${DOCKER_PULL} "
    echo " No cache during build          DOCKER_NO_CACHE             ${DOCKER_NO_CACHE} "
    echo " Use proxy during builds        DOCKER_BUILD_PROXY          ${DOCKER_BUILD_PROXY}"
    echo " Push/pull from docker hub      DOCKER_USE_HUB              ${DOCKER_USE_HUB}"
    echo " Release docker image           DOCKER_IMAGE                ${DOCKER_IMAGE}"
    echo " Use a http proxy               SET_HTTP_PROXY              ${SET_HTTP_PROXY}"
    echo " Use a pip proxy                SET_PIP_PROXY               ${SET_PIP_PROXY}"
    echo ""
    echo "Usage:"
    echo " ./develop.sh (baseimage|buildimage|devimage|releasetarball|prodimage)"
    echo " ./develop.sh (dev|dev_build)"
    echo " ./develop.sh (start_prod|prod_build)"
    echo " ./develop.sh (runtests|lettuce|selenium)"
    echo " ./develop.sh (start_test_stack|start_seleniumhub|start_seleniumtests|start_prodseleniumtests)"
    echo " ./develop.sh (pythonlint|jslint)"
    echo " ./develop.sh (ci_docker_staging|docker_staging_lettuce)"
    echo " ./develop.sh (ci_docker_login)"
    echo ""
    echo "Example, start dev with no proxy and build everything:"
    echo "SET_PIP_PROXY=0 SET_HTTP_PROXY=0 ./develop.sh dev_build"
    echo ""
    exit 1
}


info () {
  printf "\r  [ \033[00;34mINFO\033[0m ] $1\n"
}

success () {
  printf "\r\033[2K  [ \033[00;32m OK \033[0m ] $1\n"
}


fail () {
  printf "\r\033[2K  [\033[0;31mFAIL\033[0m] $1\n"
  echo ''
  exit 1
}


docker_options() {
    DOCKER_ROUTE=$(ip -4 addr show docker0 | grep -Po 'inet \K[\d.]+')
    success "Docker ip ${DOCKER_ROUTE}"

    _http_proxy
    _pip_proxy

    if [ ${DOCKER_PULL} = "1" ]; then
         DOCKER_BUILD_PULL="--pull=true"
         DOCKER_COMPOSE_BUILD_PULL="--pull"
    else
         DOCKER_BUILD_PULL="--pull=false"
         DOCKER_COMPOSE_BUILD_PULL=""
    fi

    if [ ${DOCKER_NO_CACHE} = "1" ]; then
         DOCKER_BUILD_NOCACHE="--no-cache=true"
         DOCKER_COMPOSE_BUILD_NOCACHE="--no-cache"
    else
         DOCKER_BUILD_NOCACHE="--no-cache=false"
         DOCKER_COMPOSE_BUILD_NOCACHE=""
    fi

    DOCKER_BUILD_OPTS="${DOCKER_BUILD_OPTS} ${DOCKER_BUILD_NOCACHE} ${DOCKER_BUILD_PROXY} ${DOCKER_BUILD_PULL} ${DOCKER_BUILD_PIP_PROXY}"

    # compose does not expose all docker functionality, so we can't use compose to build in all cases
    DOCKER_COMPOSE_BUILD_OPTS="${DOCKER_COMPOSE_BUILD_OPTS} ${DOCKER_COMPOSE_BUILD_NOCACHE} ${DOCKER_COMPOSE_BUILD_PULL}"
}


_http_proxy() {
    info 'http proxy'

    if [ ${SET_HTTP_PROXY} = "1" ]; then
        if [ -z ${HTTP_PROXY_HOST+x} ]; then
            HTTP_PROXY_HOST=${DOCKER_ROUTE}
        fi
        http_proxy="http://${HTTP_PROXY_HOST}:3128"
        HTTP_PROXY="http://${HTTP_PROXY_HOST}:3128"
        NO_PROXY=${HTTP_PROXY_HOST}
        no_proxy=${HTTP_PROXY_HOST}
        success "Proxy $http_proxy"
    else
        info 'Not setting http_proxy'
    fi

    export HTTP_PROXY http_proxy NO_PROXY no_proxy

    success "HTTP proxy ${HTTP_PROXY}"
}


_pip_proxy() {
    info 'pip proxy'

    # pip defaults
    PIP_INDEX_URL='https://pypi.python.org/simple'
    PIP_TRUSTED_HOST='127.0.0.1'

    if [ ${SET_PIP_PROXY} = "1" ]; then
        if [ -z ${PIP_PROXY_HOST+x} ]; then
            PIP_PROXY_HOST=${DOCKER_ROUTE}
        fi
        # use a local devpi install
        PIP_INDEX_URL="http://${PIP_PROXY_HOST}:3141/root/pypi/+simple/"
        PIP_TRUSTED_HOST="${PIP_PROXY_HOST}"
    fi

    export PIP_INDEX_URL PIP_TRUSTED_HOST

    success "Pip index url ${PIP_INDEX_URL}"
}


docker_warm_cache() {
    # attempt to warm up docker cache by pulling next_release tag
    if [ ${DOCKER_USE_HUB} = "1" ]; then
        info 'warming docker cache'
        set -x
        docker pull ${DOCKER_IMAGE}:next_release || true
        success "$(docker images | grep ${DOCKER_IMAGE} | grep next_release | sed 's/  */ /g')"
        set +x
    fi
}


# ssh setup for ci
_ci_ssh_agent() {
    info 'ci ssh config'

    ssh-agent > /tmp/agent.env.sh
    . /tmp/agent.env.sh
    success "started ssh-agent"

    # load key if defined by bamboo
    if [ -z ${bamboo_CI_SSH_KEY+x} ]; then
        info "loading default ssh keys"
        ssh-add || true
    else
        info "loading bamboo_CI_SSH_KEY ssh keys"
        ssh-add ${bamboo_CI_SSH_KEY} || true
    fi

    # some private projects had a deployment key
    if [ -f docker-build.key ]; then
        chmod 600 docker-build.key
        ssh-add docker-build.key
    fi

    ssh-add -l
}


ci_docker_login() {
    info 'Docker login'

    if [ -z ${DOCKER_USERNAME+x} ]; then
        DOCKER_USERNAME=${bamboo_DOCKER_USERNAME}
    fi
    if [ -z ${DOCKER_PASSWORD+x} ]; then
        DOCKER_PASSWORD=${bamboo_DOCKER_PASSWORD}
    fi

    if [ -z ${DOCKER_USERNAME} ] || [ -z ${DOCKER_PASSWORD} ]; then
        fail "Docker credentials not available"
    fi

    docker login -u ${DOCKER_USERNAME} --password="${DOCKER_PASSWORD}"
    success "Docker login"
}


# figure out what branch/tag we are on
git_tag() {
    info 'git tag'

    set +e
    GIT_TAG=`git describe --abbrev=0 --tags 2> /dev/null`
    set -e

    # jenksins sets BRANCH_NAME, so we use that
    # otherwise ask git
    GIT_BRANCH="${BRANCH_NAME}"
    if [ -z ${GIT_BRANCH} ]; then
        GIT_BRANCH=`git rev-parse --abbrev-ref HEAD 2> /dev/null`
    fi

    # fail when we don't know branch
    if [ "${GIT_BRANCH}" = "HEAD" ]; then
        fail 'git clone is in detached HEAD state and BRANCH_NAME not set'
    fi

    # only use tags when on master (prod) branch
    if [ "${GIT_BRANCH}" != "master" ]; then
        info 'Ignoring tags, not on master branch'
        GIT_TAG=${GIT_BRANCH}
    fi

    # if no git tag, then use branch name
    if [ -z ${GIT_TAG+x} ]; then
        info 'No git tag set, using branch name'
        GIT_TAG=${GIT_BRANCH}
    fi

    export GIT_TAG

    success "git tag: ${GIT_TAG}"
}


create_dev_image() {
    info 'create dev image'
    set -x
    docker-compose -f docker-compose-build.yml build ${DOCKER_COMPOSE_BUILD_NOCACHE} dev
    set +x
    success "$(docker images | grep muccg/${PROJECT_NAME}-dev | sed 's/  */ /g')"
}


create_build_image() {
    info 'create build image'
    set -x
    docker-compose -f docker-compose-build.yml build ${DOCKER_COMPOSE_BUILD_NOCACHE} build
    set +x
    success "$(docker images | grep muccg/${PROJECT_NAME}-build | sed 's/  */ /g')"
}


create_base_image() {
    info 'create base image'
    set -x
    docker-compose -f docker-compose-build.yml build ${DOCKER_COMPOSE_BUILD_OPTS} base
    set +x
    success "$(docker images | grep muccg/${PROJECT_NAME}-base | sed 's/  */ /g')"
}


create_prod_image() {
    info 'create prod image'
    info "Building ${PROJECT_NAME} ${GIT_TAG}"
    set -x
    docker-compose -f docker-compose-build.yml build prod
    docker tag ${DOCKER_IMAGE}:${GIT_TAG} ${DOCKER_IMAGE}:${GIT_TAG}-${DATE}
    set +x
    success "$(docker images | grep ${DOCKER_IMAGE} | grep ${GIT_TAG} | sed 's/  */ /g')"
    success 'create prod image'
}


create_release_tarball() {
    info 'create release tarball'
    mkdir -p build
    chmod o+rwx build
    set -x
    docker-compose -f docker-compose-build.yml run build
    set +x
    success "$(ls -lh build/* | grep ${GIT_TAG})"
}


start_prod() {
    info 'start prod'
    mkdir -p data/prod
    chmod o+rwx data/prod

    set -x
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-prod.yml rm --force
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-prod.yml up
    set +x
}


start_dev() {
    info 'start dev'
    mkdir -p data/dev
    chmod o+rwx data/dev

    set -x
    docker-compose --project-name ${PROJECT_NAME} up
    set +x
}


publish_docker_image() {
    # check we are on master or next_release
    if [ "${GIT_BRANCH}" = "master" ] || [ "${GIT_BRANCH}" = "next_release" ]; then
        info "publishing docker image for ${GIT_BRANCH} branch, version ${GIT_TAG}"
    else
        info "skipping publishing docker image for ${GIT_BRANCH} branch"
        return
    fi

    if [ ${DOCKER_USE_HUB} = "1" ]; then
        set -x
        docker push ${DOCKER_IMAGE}:${GIT_TAG}
        docker push ${DOCKER_IMAGE}:${GIT_TAG}-${DATE}
        set +x
        success "pushed ${tag}"
    else
        info "docker push of ${GIT_TAG} disabled by config"
    fi
}


_start_test_stack() {
    info 'test stack up'
    mkdir -p data/tests
    chmod o+rwx data/tests

    set -x
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-teststack.yml up $@
    set +x
    success 'test stack up'
}


_stop_test_stack() {
    info 'test stack down'
    set -x
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-teststack.yml stop
    set +x
    success 'test stack down'
}


start_test_stack() {
    _start_test_stack --force-recreate
}


run_unit_tests() {
    info 'run unit tests'
    _start_test_stack --force-recreate -d

    set +e
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-unittests.yml run --rm testhost
    rval=$?
    set -e

    _stop_test_stack

    return $rval
}


_start_selenium() {
    info 'selenium stack up'
    mkdir -p data/selenium
    chmod o+rwx data/selenium

    set -x
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-selenium.yml up $@
    set +x
    success 'selenium stack up'
}


_stop_selenium() {
    info 'selenium stack down'
    set -x
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-selenium.yml stop
    set +x
    success 'selenium stack down'
}


start_seleniumhub() {
    _start_selenium --force-recreate
}


start_lettucetests() {
    set -x
    set +e
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-lettuce.yml up --force-recreate
    local rval=$?
    set -e
    set +x

    return $rval
}


lettuce() {
    info 'lettuce'
    _start_selenium --force-recreate -d
    _start_test_stack --force-recreate -d

    start_lettucetests
    local rval=$?

    _stop_test_stack
    _stop_selenium

    exit $rval
}


start_seleniumtests() {
    set -x
    set +e
    docker-compose --project-name ${PROJECT_NAME} -f docker-compose-seleniumtests.yml up --force-recreate
    local rval=$?
    set -e
    set +x

    return $rval
}


selenium() {
    info 'selenium'
    _start_selenium --force-recreate -d
    _start_test_stack --force-recreate -d

    start_seleniumtests
    local rval=$?

    _stop_test_stack
    _stop_selenium

    exit $rval
}
